process/peer/process.go

```diff
@@ -327,6 +359,90 @@ func (p *validatorStatistics) getPeerAccount(address []byte) (state.PeerAccountH
 	return peerAccount, nil
 }
 
+// loadPreviousShardHeaders loads the previous shard headers for a given metablock. For the metachain it's easy
+//  since it has all the shard headers in it's storage, but for the shard it's a bit trickier and we need
+//  to iterate through past metachain headers until we find all the ShardData's we are interested in
+func (p *validatorStatistics) loadPreviousShardHeaders(currentHeader, previousHeader *block.MetaBlock) error {
+
+	if p.shardCoordinator.SelfId() > p.shardCoordinator.NumberOfShards() {
```

Comment:
2 implementations for this, one for shard and one for meta. it can be decided at peer Processor creation.